crates/common/tedge_config_macros/examples/macro.rs

```diff
@@ -0,0 +1,178 @@
+                #[doku(as = "PathBuf")]
+                #[serde(alias = "capath")]
+                ca_path: Utf8PathBuf,
+
+                /// Path to the client certficate
+                #[doku(as = "PathBuf")]
+                #[serde(alias = "certfile")]
+                cert_file: Utf8PathBuf,
+
+                /// Path to the client private key
+                #[doku(as = "PathBuf")]
+                #[serde(alias = "keyfile")]
+                key_file: Utf8PathBuf,
+            }
+        },
+
+        external: {
+            bind: {
+                /// The port mosquitto binds to for external use
+                #[tedge_config(example = "8883")]
+                port: u16,
+
+                /// The address mosquitto binds to for external use
+                #[tedge_config(example = "0.0.0.0")]
+                address: IpAddr,
+
+                /// Name of the network interface which mosquitto limits incoming connections on
+                #[tedge_config(example = "wlan0")]
+                interface: String,
+            },
+
+            /// Path to a file containing the PEM encoded CA certificates that are
+            /// trusted when checking incoming client certificates
+            #[tedge_config(example = "/etc/ssl/certs")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "capath")]
+            ca_path: Utf8PathBuf,
+
+            /// Path to the certificate file which is used by the external MQTT listener
+            #[tedge_config(note = "This setting shall be used together with `mqtt.external.key_file` for external connections.")]
+            #[tedge_config(example = "/etc/tedge/device-certs/tedge-certificate.pem")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "certfile")]
+            cert_file: Utf8PathBuf,
+
+            /// Path to the key file which is used by the external MQTT listener
+            #[tedge_config(note = "This setting shall be used together with `mqtt.external.cert_file` for external connections.")]
+            #[tedge_config(example = "/etc/tedge/device-certs/tedge-private-key.pem")]
+            #[doku(as = "PathBuf")]
+            #[serde(alias = "keyfile")]
+            key_file: Utf8PathBuf,
+        }
+    }
+}
+
+fn device_id(_reader: &TEdgeConfigReader) -> Result<String, String> {
+    Ok("dummy-device-id".to_owned())
+}
+
+fn default_device_key(location: &TEdgeConfigLocation) -> Utf8PathBuf {
+    location
+        .tedge_config_root_path()
+        .join("device-certs")
+        .join("tedge-private-key.pem")
+}
+
+fn default_device_cert(location: &TEdgeConfigLocation) -> Utf8PathBuf {
+    location
+        .tedge_config_root_path()
+        .join("device-certs")
+        .join("tedge-certificate.pem")
+}
+
+fn default_mqtt_port() -> NonZeroU16 {
+    NonZeroU16::try_from(1883).unwrap()
+}
+
+fn main() {
+    let dto = TEdgeConfigDto::default();
+    let config = TEdgeConfigReader::from_dto(&dto, &TEdgeConfigLocation);
```

Comment:
I would add this to the example, as this is that *straight* use the generated structs that let me prefer this proposal.

```suggestion
    let mut dto = TEdgeConfigDto::default();
    dto.mqtt.bind.address = Some(IpAddr::V4(Ipv4Addr::new(1, 2, 3, 4)));

    let config = TEdgeConfigReader::from_dto(&dto, &TEdgeConfigLocation);
    let host: IpAddress = config.mqtt.bind.address;
    let port: NonZeroU16 = config.mqtt.bind.port;
    println!("mqtt = {}:{}", host, port);

```

Reply:
Wouldn't lines 172 and 173 move values out of config? Is this intended?

Reply:
> Wouldn't lines 172 and 173 move values out of config? Is this intended?

No, because `IpAddress` and `NonZeroU16` both implement `Copy`.